### PR TITLE
Indentation cleanup, added smart_brackets and smart_insertions, and more

### PR DESCRIPTION
**Pull request by eidheim**
@d3rrial, @zalox and others: I need testing of this PR before merge:
* Added move one ) instead of insert, which is very nice after autocomplete (#281)
* Major cleanup of the indenting code, it's actually much faster compared to current master, and additionally more readable (I hope). 
* smart_insertions
* Various other improvements

### CODE REVIEW

**Comment by eidheim** [comment]
And the indentation is improved as well. For instance it should now correctly indent:
```c++
some_function(true, 42,
              [] {//enter here
```

, that will lead to:

```c++
some_function(true, 42,
              [] {
  //cursor here
}
```

**Comment by d3rrial** [comment]
```
void asdf(std::string& superduperlongvariablenamewtf, std::string& andanotheronewhowouldbelievethis,
          std::string& thisiswaytoomuch){
            
          }
```
Is the indent here as intended? Looks strange to me.

Other than that it looks good to me.

**Comment by eidheim** [comment]
@d3rrial Thank you, that is definitely a bug. I'll fix it Tomorrow.

**Comment by eidheim** [comment]
I fixed the bug @d3rrial mentioned. Also improved indentation when writing constructor implementations on multiple lines. Let me know if there are other issues:

**Comment by eidheim** [comment]
See also the last. You can now avoid using arrow keys when for instance writing:
```c++
auto c=std::make_shared<int>(10);
```

File: src/source.h [hunk 1]
```diff
@@ -124,9 +124,10 @@ namespace Source {
     
     Gtk::TextIter find_start_of_sentence(Gtk::TextIter iter);
     bool find_open_non_curly_bracket_backward(Gtk::TextIter iter, Gtk::TextIter &found_iter);
+    bool find_close_non_curly_bracket_forward(Gtk::TextIter iter, Gtk::TextIter &found_iter);
     bool find_open_curly_bracket_backward(Gtk::TextIter iter, Gtk::TextIter &found_iter);
     bool find_close_curly_bracket_forward(Gtk::TextIter iter, Gtk::TextIter &found_iter);
-    long open_close_paranthesis_count(Gtk::TextIter iter);
+    long open_close_bracket_count(Gtk::TextIter iter, unsigned int left_bracket, unsigned int right_bracket);
```
Comment (d3rrial) [inline src/source.h]: `unsigned int`? Are you sure? Wouldn't that mean the `'('` has to be cast from char? Or does Clang handle single quoted symbols differently?

Comment (eidheim) [inline src/source.h]: `*iter` returns `unsigned int` since it is a UTF-8 character. So I do the cast automatically in the function parameters. At least the tests passed:)

**Comment by d3rrial** [comment]
`auto c=std::make_shared<int>(10);`
Works well, but not flawlessly (this might be out of scope for this PR, tho):

when I type auto `c=std::make_sha` I get an autocomplete suggestion for

`auto c=std::make_shared<typename _Tp>(_Args &&__args...)`

where my cursor sits in the angled bracketes with `typename_Tp` being highlighted (so it's overwritten when I start typing). To get past that initial `>` after the `typename_Tp` I still have to use the arrow key to skip it, tho. Other than that, it works perfectly.

File: src/source.h (hunk 1 above)
Comment (d3rrial) [inline src/source.h]: Well it wasn't an error, at most I would have expected the compiler to throw a warning. As long as it works, it's all good :)

**Comment by eidheim** [comment]
In the case of `auto c=std::make_shared<typename _Tp>(_Args &&__args...)` and you replace typename _Tp you should get moved when entering `>` just before the `>`. You should be able to enter both template argument and function argument without using the keys after using autocompletion on std::make_shared.

**Comment by d3rrial** [comment]
This doesn't work for me. Could it be because of different keyboard layouts / keyval differences?

**Comment by eidheim** [comment]
Yeah, tested on Linux. Have to find the right key that works (GDK_KEY_greater works on OS X).

**Comment by d3rrial** [comment]
I just tested and GDK_Key_greater is keycode 62, but 62 is also they keyval for > for me, so that's probably not the cause...

**Comment by eidheim** [comment]
The problem is searching for `<` and `>` keys it seems. I'll see if I can fix it later.

**Comment by d3rrial** [comment]
Well, `<` and `>` are also comparison operators after all. In the case where it's used as a templating mechanism, I think it's enough to stop as soon as any other type of bracket is found. I can't think of a situation where `<...)>` or `<...]>` or `<...}>` would be valid constructs.

```cpp
auto a=crazytemplate<std::string, void*, int,
                     double, size_t>(...);
```
This has to remain valid, tho...

**Comment by d3rrial** [comment]
Looks good to me, even works with nested templates. Haven't done any edge-case tests for false positives yet, but I haven't had any false negatives. :+1:

**Comment by eidheim** [comment]
Thank you, good to hear. At least there is some progress:)

**Comment by eidheim** [comment]
I made the rules for the movements very strict so that they will not happen unless you want them to, hopefully.

**Comment by d3rrial** [comment]
While I personally love this feature / these movements and will/would use them all the time, for some people this might be an inconvenience (I can't fathom how, because they behave just like regular, solid brackets in any case other than when you type in a bracket which is already there) so I think it might be best if this was put into a setting in the config file.

**Comment by eidheim** [comment]
I'm thinking that too. I'll use it myself, but we should maybe have it disabled by default. What do you think @zalox ?

**Comment by zalox** [comment]
See  https://github.com/cppit/jucipp/issues/281#issuecomment-244985254.

I see no reason to not enable them by default. It would effectively improve the coding experience for new users. People like me, that don't like these features, know how to disable them and keep quiet as long as there is an option to do so.

**Comment by eidheim** [comment]
So what should we call this bracket feature in the preferences? `smart_brackets`? I'll add some comment as well (`smart_brackets_comment`).

**Comment by eidheim** [comment]
Not completely sure about the preference item name `smart_brackets`, but at least the functionality is there.

**Comment by eidheim** [comment]
I went a bit crazy and implemented "smart insertations". Needs testing!:)

**Comment by eidheim** [comment]
Added some improvements, and decided to drop smart insertations on `{}`; it was just too annoying at least the way I work:)

**Comment by d3rrial** [comment]
Works perfectly in the few scenarios I came up with. I'm going to test this more long term to be sure, but from my perspective this looks good!

**Comment by zalox** [comment]
I'm having a look tonight :) @tedjk @oyvang You should also test juCi++ now.

**Comment by eidheim** [comment]
It's not 100% perfect yet, and one important rule I'll add is that one should never add pairs when in front of a letter, underscore or a number. For instance:
`if(|true) //cursor at |`
should become
`if((|true)` instead of `if((|)true)`

There are probably more such rules as well. 

There is also some minor issue when the cursor is in front of `"something"` or `'s'`.

**Comment by eidheim** [comment]
Also, I guess that it does not make sense to add `()` in front of a `(`. The result should be `((...)` instead of `()(...)` I take it?

**Comment by d3rrial** [comment]
Well, unless you want to C-cast the result of a calculation

**Comment by eidheim** [comment]
Good point!

**Comment by eidheim** [comment]
But still I think it is more annoying that `()` comes up when you want `(`, than the other way around. Though, this is a personal preference. We should try keep the false positives to a minimum though I belive. But would love to hear your opinions on this.

**Comment by d3rrial** [comment]
I totally agree with you on that. C-style casts shouldn't be used in C++ anymore anyway, so this is not relevant very often and not having the bracket appear does not cause any inconvenience I believe.

Btw. do you have a keyboard shortcut to surround a given highlighted text with brackets?

**Comment by eidheim** [comment]
I know gtk has some such functions that one can create shortcuts for (in a gtk-init file), but have not looked into this much.

**Comment by eidheim** [comment]
A large cleanup again, also activated smart brackets/insertions for all files, and added exceptions for () and [] insertions.

**Comment by eidheim** [comment]
Fixed various smaller issues in last commit.

**Comment by d3rrial** [comment]
I've found one thing that I dislike:
If I type
```cpp
void a
```
and open the bracket, it autocompletes to:
```cpp
void a()
```
if I (for some reason) don't want the second bracket tho and hit "Delete" to delete the character to the right of the cursor, both brackets get deleted and I'm back with
```cpp
void a
```
but unable to open just a single bracket. This is resolved by typing a single letter within the brackets, but personally I'd prefer it if I could hit the delete key and delete only characters to the right of the cursor.

**Comment by eidheim** [comment]
@d3rrial fixed in last commit, thank you!

**Comment by eidheim** [comment]
Added some more fixes, getting closer to the finishing line:) We should probably test this for some extra days though. Let me know if there are more issues:)

**Comment by eidheim** [comment]
Ok, I'm happy with this PR, at least right this minute. It should be pretty optimised too. We should test it some more days though. And if there is anything that is annoying, even the slightest, let me know:)

**Comment by d3rrial** [comment]
I'm a little sad that this doesn't seem to work with other languages (for example Rust) for which I'm also using juCi++. Maybe allow specifically smart `""` and `()` insertion in other languages as well? I can't think of a high level language where the insertion rules for strings or brackets would be different from the syntax C++ uses. (But my knowledge of languages is also quite limited)

**Comment by eidheim** [comment]
It should be enabled for all languages, but it depends on how good the language definition files are for gtksourceview. You can use it in CMakeLists.txt for instance, but cmake's language definition files are missing a string definition (it still works though).

**Comment by d3rrial** [comment]
Oh, OK. Strange. I would've thought it would work then... Rust has complete syntax highlighting and keyword autocomplete. I must've overestimated how good the language definition files are, then. Thanks!

**Comment by eidheim** [comment]
See the language definitions in `/usr/share/gtksourceview-3.0/language-specs/`. For instance python.lang has <context ... class=...>, but the rust.lang file is missing this attributes. You can add them your self though. juCi++ looks for class or class-disabled attributes in the context nodes, and if found enables for instance smart_insertions to work. See https://github.com/cppit/jucipp/blob/master/src/source.cc#L1853.

**Comment by d3rrial** [comment]
Well, since it's obviously working from your side but not from theirs, I'll just wait for better Rust integration into gtksourceview. Thanks again! :)

**Comment by eidheim** [comment]
I think I fixed it. It was basically my bad.

**Comment by d3rrial** [comment]
And now I sound passive aggressive... :P
Thanks for the fix :)

**Comment by eidheim** [comment]
Just added smart insertions on selected text.

**Comment by d3rrial** [comment]
I like it.
I'm still internally debating if it would be better in its own setting or if it's good like it is. On one hand it's really useful when you want to surround something with brackets, on the other hand it overrides the common notion that typing a character while having something highlighted will replace the highlighted text.

I'm personally for leaving it as it is.
I think it when the time comes for documention this should be properly explained, tho.

**Comment by eidheim** [comment]
We have some freedom when a user activates this option I think. Also added insertions of
```
`
```
(dead grave) for markdown on selected text.

**Comment by eidheim** [comment]
I'm going to rename `smart_brackets` to `smart_moves` I think (in preferences and elsewhere). Does that make sense? Maybe rename `smart_insertions` to `smart_inserts` too? Not sure about the latter one though.

**Comment by d3rrial** [comment]
I think `smart_inserts` is better than `smart_insertions`, but I like `smart_brackets` more than `smart_moves`
smart_moves seems so nondescript if you don't know the context.

**Comment by eidheim** [comment]
Let's just keep it as it is then, easiest solution anyway since I can't decide:) AutoPairs does not include inserting `;` for instance, so it is not descriptive enough I think.

**Comment by eidheim** [comment]
I also did a spellcheck cleanup now that we have a working is_code_iter! The spellcheck code should be more understandable now.

**Comment by eidheim** [comment]
Ah, and testing on the new spellcheck code is needed as well:)

**Comment by d3rrial** [comment]
That reminds me, is there a way to disable the spell-check entirely? It often flags my variable names for silly reasons. It's not that important since it has no influence on my workflow, it's just a visual annoyance.

**Comment by eidheim** [comment]
You can set the spellcheck language to some invalid value in preferences. However, variable names outside strings and comments should not be spellchecked.

**Comment by d3rrial** [comment]
Since the latest commit or before that already?
I think I had the spellchecker tell me off for variable names, but it could've also been strings, I'm tired right now and I can't trust my memories. I'll test it again tomorrow and report back.

**Comment by eidheim** [comment]
The spellchecker should be context aware, and has been since the beginning, but there are some occasions where it fails to underline properly like when doing undo/redo/pasting, but it should not underline code at all. If it does, it is removed after a short delay.

**Comment by eidheim** [comment]
Fixed the spellcheck issue on undo/redo/pasting.

**Comment by zalox** [comment]
Shouldn't the `{` also be smart_inserted?

**Comment by eidheim** [comment]
I tried doing smart insertion of {}, but found that too often that was not what I wanted. Additionally, we already have smart indentation and insertion of } after pressing enter after {, and this leads to far less false positives.

**Comment by eidheim** [comment]
@zalox This PR should be getting ready for merge. I added some more improvements that should be pretty safe. The biggest change is the indentation cleanup, and that seems to work fine.

**Comment by eidheim** [comment]
@zalox this PR should now be ready for merge!:)

**Comment by eidheim** [comment]
I'll merge this now as this PR is stable and contains numerous improvements.
